### ISSUE TITLE
Create example, and ensure that 2 nodes can coexist on 1 Python VM

### ISSUE THREAD

**Issue by kvakvs** [Pyrlang/Pyrlang#43]

**Comment by rlouf** [comment]
Are you currently working on this?

**Comment by kvakvs** [comment]
Yes but you can give it a try if you wish. I have changed the description.

I am using branch which been never pushed yet, cannot promise huge progress this week.

**Comment by s2hc-johan** [comment]
I've started to work on this now. I've rebased the branch onto master and pushed it to my repo.

https://github.com/s2hc-johan/Pyrlang/tree/pyrlang2-asyncio

I think I've managed to get the examples running. But I don't really know exactly what the last ones should do so it's hard to know if it's working or not.

Some things came to mind that I though you might help me with @kvakvs 
* Signals for processes are round tripping into the node state, is that a necessity or legacy from
  previous design. if not we could clean up a bit there.
* inbox_ on the node class, is that needed, or can we handle that functionality with asyncio create task? or are there timing things that it needs to be handled in order?
* The internal functions send, link, monitor etc which ones are blocking and not. I need to know so
  that we can decide which ones we can wrap so they can be called from non `async` functions

**Comment by kvakvs** [comment]
> Signals for processes are round tripping into the node state
I believe that is not necessary, you can simplify
> inbox_ on the node class, is that needed
There was some consideration early, but i don't remeber what a node can receive, safe to not implement and see if you ever need it later
> The internal functions send, link, monitor 
Local versions are non-blocking, instant effect.  Remote versions might need to send something to the remote. See the actual dist protocol which of these functions trigger a message, those become async.

**Comment by s2hc-johan** [comment]
@kvakvs thanks for the clarification

Ok, so everything that uses network or similar is going `async` however what I thought of was how we're going to make the interface of the node. If it's ok not to wait for the completion of a link when it's remote, I think the public api should not be a coroutine. I'm thinking something like this:

```
def send(self, sender, receiver, message) -> None:
    send_task = self._send(sender, receiver, message)
    asyncio.get_running_loop().create_task(send_task)

async _send(self, sender, receiver, message) -> None:
    ...

async def link(self, pid1, pid2, local_only=False):
    if is_pid_local(...):
        self._local_link(...)
    else:
        link_task = self._remote_link(...)
        async.get_running_loop().create_task(link_task)

async def _remote_link(self, pid1, pid2, local_only=False):
    ...
```

The good thing is that we can call it from functions in dist protocol, since they're not coroutines. The bad thing is if we can't wait for them in a coroutine with `await node.send(...)`. I'm leaning towards this being the neater interface, and we could always make a `coro_send` or similar if the need would show itself.

**Comment by kvakvs** [comment]
You can consider that sometimes you will be needing the result, and sometimes you'll be inside an async so await-ing is not a problem.
I suggest for each async call you create a mirror normal call which will call the async, wait for it  to complete and return the result. Something like `Future().set_result(x)` (not sure what's the correct way)
so it will be something like `def link_sync`, and `async def link`

**Comment by kvakvs** [comment]
You can consider that sometimes you will be needing the result, and
sometimes you'll be inside an async so await-ing is not a problem.
I suggest for each async call you create a mirror normal call which will
call the async, wait for it  to complete and return the result. Something
like `Future().set_result(x)` (not sure what's the correct way)
so it will be something like `def link_sync`, and `async def link`

**Comment by s2hc-johan** [comment]
@kvakvs I’m following the _nowait nomenclature Asyncio uses. 

How about passive mode for process. Is that something you made for CPU considerations or is it mirroring a function in Erlang. Now with full Asyncio, maybe we could incorporate hibernation?

**Comment by s2hc-johan** [comment]
@kvakvs As I'm working with process and gen server one thing came to mind. The loop thing we do in process, even though that is what many people will want, a process doesn't have a loop in erlang right.

I'm done with a base of receive functionality, you already done the inbox thing. What about only implementing some form of state in the process like `['init', 'running', 'exiting', 'done']` and then we implement a run function.

The genserver will be as simple if not better I think and we can remove the `passive` functionality from process, we can always write a small class `ReactiveProcess` that implements a runfunction that waits for inbox args and run something, if people don't want the genserver class

**Comment by kvakvs** [comment]
I think best way to make process do infinite receive would be having one `asyncio.Queue` and there you `await` for a value from that queue, in a separate task. For selective receive it is more complicated you will need 2 queues, one for newly arrived, and one for checked but not received. Not sure whether you need that.

**Comment by s2hc-johan** [comment]
I'm already done with that, testing it out right now. The thing was to remvoe the infinit receive loop in process, since that isn't part of an erlang process right.

**Comment by kvakvs** [comment]
Erlang mailbox does receive everything automatically (like `asyncio.Queue` does) and actual receive is happening when the user logic needs it. There is no receive loop, but in gen_server/gen* processes there is a loop indeed.

**Comment by s2hc-johan** [comment]
exactly so don't we mimic it better with a `run` function in process, you can build whatever logic you want, calling receive as you like.

Then in gen_server we implement a loop, just like in the erlang case

**Comment by s2hc-johan** [comment]
Another thing I just ran into, when using `node.send` we want the destination to be `(node, process)` but when we send from erlang we use `{Process, Node)` isn't it better to use the erlang way when doing erlang communication?

**Comment by kvakvs** [comment]
Yes you're right, erlang-way sounds good. And prefer to use keyword arg names everywhere, in examples too: `Node.send(node=X, process=Y)`

**Comment by s2hc-johan** [comment]
Nice I'll try to fix that.

I'm somewhat done with base parts in my repo for asyncio https://github.com/s2hc-johan/Pyrlang/commits/pyrlang2-asyncio

Feels quite performant, almost no sleeps (got an idea to remove them all soon), and I can make gen_server calls and cast both from erlang, python and between python nodes.

do you want me to create a PR to discuss there, or should we continue discussion here, and you'll just look at my repo.

Some issues I've found

- [ ] if we try to initiate node connection between 2 pyrlang nodes at the same time it hangs, probably need to implement the `ok_simultaneous` response in handshake
- [ ] if I leave example8 (the py nodes) open it starts bursting data between them.

Then there is improvement around the pattern matching and the decorators, but that could be evolved a bit later.

**Comment by kvakvs** [comment]
Put a log error on the ok_simultaneous, and you will know whether it is that situation happening.
We send empty command as a ping, and we reply with empty command as a response, they might just trigger mutually each other and create an endless wave of pings and replies to replies to replies... Can put a timer limit there maybe, to not respond to the same node more often than every 10-30sec? Or modify the ping packet to know if its a request or response.

**Comment by s2hc-johan** [comment]
is the ping a pyrlang thing or an erlang thing. If erlang thing how does the erlang nodes handle this?

**Comment by kvakvs** [comment]
Not sure, the documentation doesn't say how to handle that.
First try and add log to the protocol, and discover what is that traffic, is that really pings? And then we figure something out.

**Comment by s2hc-johan** [comment]
so I found that we're blindly responding to a ping
https://github.com/s2hc-johan/Pyrlang/blob/pyrlang2-asyncio/pyrlang2/dist_proto/base_dist_protocol.py#L376

so that becomes a endless loop. There is really no reason to respond to that since both client and server will send out a ping. we should instead have another timed event, that checks that we've gotten some form of interaction from the node in the last 15s

**Comment by kvakvs** [comment]
Yes, the good solution i reckon would be to remember for each node when a last message from it was received, and when a last message to it was sent. And if the difference between these two times becomes too big, we time out and close.

**Comment by s2hc-johan** [comment]
So I've updated the branch. looks ok. One problem that I found is that monitor is not async so when I try to monitor a process and it does a `dist_command` that is created as a task, so we end up with two connections to the other node. If we add the `dist` object to the node immediately and instead poll for it to reach connected state, we could prevent two connections being opened.
